@damassi => [PageView Tracking] Correct for over-tracking on tabbed routes

mzikherman opened the pull request:
You called it in https://github.com/artsy/reaction/pull/3076#discussion_r362671969 .

This wound up overtriggering on tabbed pages.

This additional tweaked check seemed to work as desired in my local testing, but now I'm wondering if we should revert this approach entirely (and leave things on `componentDidMount`).

That would definitely remove over-triggering, but it puts us back in the position we were for the A/B test, where we were dropping all those subsequent calls. My gut tells me we have to get these lifecycle/hook calls correct in order to capture all those analytics properly.

joeyAghion commented:
Good call @damassi and thanks for your continued follow-up @mzikherman.

I tend to agree that it will be important to get this operating correctly within Reaction at some point, if not now. The work you do here to compare the page's parameters seems intuitively correct, but I don't have much real insight.

Is it possible to decide tracking based on the level in the page/component hierarchy that's being updated? I.e., if the whole page needs re-rendering, automatically consider it a full page-view, and _don't_ for re-renders of components nested within the page.

mzikherman commented:
> I.e., if the whole page needs re-rendering, automatically consider it a full page-view, and don't for re-renders of components nested within the page.

That's already violated with some of our tabbed pages. For instance, switching to another tab on the artist page doesn't re-render the whole page (just the tabbed contents). And switching to other pages in client side routing won't have whole page reloads. Meanwhile, interacting with the filter _does_ update the URL, but naturally _shouldn't_ be considered a pageview.

It's very tricky! There's lots of cases to get right...

mzikherman commented on src/Artsy/Analytics/trackPageViewWrapper.tsx:
```diff
@@ -1,19 +1,47 @@
+      }
+    }
+
+    // TODO: Type props properly, can have optional `match`
```
I spent a while trying to get the generic props to extend an interface that optionally included a `match` property, but couldn't quite get it. Since we're using `get` anyway, I just rolled with it.

damassi commented:
@mzikherman - While unerelated to this case I would recommend reviewing the links I posted in the React docs regarding `React.Memo` and `React.PureComponent`, for future reference. The idea is to implement those higher-level constructs which will intelligently determine whether a component should be refreshed by internally checking whether the previous props or state are different than the current state. For components with props that aren't complex, it's usually a one-liner as it will do a shallow check; this also improves overall performance. For deeply nested props one often needs to write a more sophisticated comparitor function via something that deeply checks objects (like `isEqual` from lodash, like you're using here.) 

For this problem it might be worthwhile to invesitgate the [`onChange` callback in the `<Tabs>` component](https://palette.artsy.net/elements/navigation/tabs/) and handle tracking there. Top level tracking would be handled route level, from the page shell, then subsequent calls can be made from inside of this function and handled appropriately -- eg 

```js
if (alreadyTracked(tab)) {
  return 
} else {
  trackEvent(...)
}
```

mzikherman commented:
I'm going to give this another shot today. I don't fully like this solution, and I'm now realizing that the tab navigation _still_ relies on mediator and Force-side analytics (😢 )

Will try to remove those and re-try this!

mzikherman commented:
Pushed an update that will require a Force update to bump Reaction properly, removing some mediator hooks.

The essential change is removing the `componentDidUpdate` code entirely, that feels unsafe to really use critically. Too many moving parts, re-renders, etc.

Replacing that are simply more explicit `onClick` props to trigger pageviews where appropriate. 

I added them to the tabs (which replaces the Force code in https://github.com/artsy/force/blob/e6f923d67a90fa866022288356508ca6a2602236/src/desktop/apps/artist/client.tsx#L34), and also added them to `RouterLink`, which should solve the originally reported bug of dropping pageview events when navigating from one collection to another, currently.

I still have to see how this works with our full client side routing, but since that's not out yet in production and this does fix bugs currently out in production (and in a much more reasonable way than `componentDidUpdate` IMO), this is safe to review and merge.

In the meantime, I'll see how this feels with full client side routing.

mzikherman commented:
I think that with full client side routing, while solving the outstanding bugs, this _does_ bring us back to where we were: namely, we're not capturing subsequent pageviews when you stay on one app (clicking from one artist to another).

Those wind up relying on link interception, so we may be able to be clever there (comparing locations, etc.). Either way that can be a separate task.

mzikherman commented on src/Artsy/Router/RouterLink.tsx:
```diff
@@ -41,8 +42,22 @@ export const RouterLink: React.FC<LinkProps> = ({ to, children, ...props }) => {
+    // those components will get mounted with their own tracking.
+    // But when linking within the same app (just different params),
+    // we want to trigger a pageview.
+    const onClick = e => {
```
So..yea. Finding that when using this to navigate from a collection to another collection, we want to trigger the pageview.

But, when navigating from an artist page to a collection, we'd want to defer to the regular tracking happening on mount in those apps.

This so far is working (in the current production setting of just having client side routing in the collect suite of apps), as well as with the flag for global client side routing turned on (subsequent pageviews in the same type _are_ triggered).

damassi replied:
#minor `e` -> `event`

damassi commented on src/Artsy/Router/RouterLink.tsx:
```diff
@@ -41,8 +42,22 @@ export const RouterLink: React.FC<LinkProps> = ({ to, children, ...props }) => {
+    // we want to trigger a pageview.
+    const onClick = e => {
+      const toPath = to.toString()
+      const currentPageType = window.location.pathname.split("/")[1]
```
Can you pluck `location.pathname` from the router stored in our `SystemContext` via `useSystemContext`? We should avoid the mixing window with router instance state if possible.

mzikherman replied:
Will look into that!

damassi replied:
Looking at the file I just noticed [this too](https://github.com/artsy/reaction/blob/833168be194c3ed855c5953a64594e69b8e479ad/src/Artsy/Router/RouterLink.tsx#L19) -- so maybe that's already in scope?

mzikherman replied:
Ah yes, totally.

damassi commented on src/Artsy/Router/RouterLink.tsx:
```diff
@@ -41,8 +42,22 @@ export const RouterLink: React.FC<LinkProps> = ({ to, children, ...props }) => {
+      const currentPageType = window.location.pathname.split("/")[1]
+      const toPageType = toPath.split("/")[1]
+
+      if (currentPageType === toPageType) trackPageView({ path: toPath })
```
Another minor nit (feel free to ignore if you feel strongly otherwise) 

```js
if (currentPageType === toPageType) {
  trackPageView({ path: toPath })
}

if (props.onClick) {
  props.onClick(event)
}
```
one-liner if's are prone to bugs over time and are harder to read than common multiline ifs.

mzikherman replied:
Cool, yea I like this better.

damassi commented on src/Artsy/Router/RouterLink.tsx:
```diff
@@ -41,8 +42,27 @@ export const RouterLink: React.FC<LinkProps> = ({ to, children, ...props }) => {
+    // those components will get mounted with their own tracking.
+    // But when linking within the same app (just different params),
+    // we want to trigger a pageview.
+    const onClick = event => {
```
Sorry! one more small one -- handlers are conventionally `handle<Event>`

mzikherman commented:
Updated with specs. Thanks for the thorough review!

artsyit commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in v24.0.21 :rocket:
